[DSM] PEPPER-754/PEPPER-840 Juniper Kit classes, unit test and shipKit endpoint

pegahtah opened the pull request:
## Context


PEPPER-840
This PR is now updated to show changes for  PEPPER-840, it includes the endpoint and the service needed to create the kit

-----------
PEPPER-754
I created a unit test to mock what Juniper is supposed to send us in JuniperKitIntegration.java test class and created the JuniperShipKitRoute.java class basically from copied code to make sure the kit can be created in the System.

Little bit of tweaks here and there to make the code functional 
 

## Checklist

- x ] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [x] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [x] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

pegahtah committed:
PEPPER-754
I created a unit test to mock what Juniper is supposed to send us in JuniperKitIntegration.java test class and created the JuniperShipKitRoute.java class basically from copied code to make sure the kit can be created in the System.

Little bit of tweaks here and there to make the code functional

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-754

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1184,8 +1184,9 @@ public static void updateKit(String dsmKitId, Shipment participantShipment, Ship
         } else {
-            logger.info("Updated kit w/ dsm_kit_id " + dsmKitId, results.resultException);
-
+            logger.info("Updated kit w/ dsm_kit_id " + dsmKitId);
```
I fixed the error and move the upsert into a check

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/ddp/instance/DDPInstanceDao.java:
```diff
@@ -264,6 +265,30 @@ public Optional<DDPInstanceDto> getDDPInstanceByInstanceName(@NonNull String ins
         return Optional.ofNullable((DDPInstanceDto) results.resultValue);
     }
 
+    public Optional<DDPInstanceDto> getDDPInstanceByStudyGuid(@NonNull String studyGuid) {
```
This method was missing in DDPInstanceDao, basically this is to be able to select the ddpInstaceDto based on the stuyd-guid

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,320 @@
+import spark.Request;
+import spark.Response;
+
+public class JuniperShipKitRoute extends RequestHandler {
```
This route at this point has many assumptions and unanswered questions, basically I have moved code from other parts of DSM and changed them to fit Juniper, but kept a commented block here and there in case the assumptions are changed in the next days

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,320 @@
+//        }
+    }
+
+    private void handleKit(@NonNull Connection conn, @NonNull DDPInstance ddpInstance, @NonNull KitType kitType, @NonNull KitRequest kit,
```
When it is more figured out of what is and is not needed for Juniper, we can merge some of these handleKit methods into one

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/JuniperKitRequest.java:
```diff
@@ -0,0 +1,22 @@
+    private String postalCode;
+    private String country;
+    private String phoneNumber;
+    String juniperKitId;
```
Could re rename `juniper*` to `external*`?  Today it's juniper, but the API could also be used by other systems.

zyme replied:
Gah never mind, I see we have `external` all over the place already...

pegahtah replied:
yeah... GBF...

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1184,8 +1184,9 @@ public static void updateKit(String dsmKitId, Shipment participantShipment, Ship
-            logger.info("Updated kit w/ dsm_kit_id " + dsmKitId, results.resultException);
-
+            logger.info("Updated kit w/ dsm_kit_id " + dsmKitId);
+        }
```
So there's a case where the `setDsmKitId` call fails but we still want to continue?
If so, maybe an explanation?

pegahtah replied:
Oh I see what you mean, great catch! I will fix this

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,320 @@
+        return null;
+    }
+
+    public Object createAJuniperKit(List<KitRequest> kitUploadObjects, String studyGuid, String kitTypeName,
```
To echo Zim's point elsewhere, would createKit() work here instead?
Improves flexibility in case you want to use this more broadly, and I assume there is nothing particular to Juniper in this method?

pegahtah replied:
that makes sense, 
From my view, we are building a way for DSM to handle none-Pepper kits, so maybe we need some kind of a name like `createANonPepperKit()` ?

denniscunningham replied:
Perhaps a nit, but articles (the, a) are pretty much never used when naming things (at least in the software I've seen). So perhaps `createNonPepperKit`?

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperKits/JuniperKitIntegration.java:
```diff
@@ -0,0 +1,61 @@
+import org.junit.Ignore;
+import org.junit.Test;
+
+public class JuniperKitIntegration {
```
Seems like our convention is to include the term `Test` on all tests?
A nit, but I'm not sure the term `Integration` adds much here since we typically don't label tests that way?

pegahtah replied:
Good point, how about JuniperKitCreationTest?

pegahtah committed:
Merge branch 'develop' into PEPPER-754

pegahtah committed:
PEPPER-754
PR comments and some more fixes and removing special kits

pegahtah committed:
Merge branch 'PEPPER-754' of github.com:broadinstitute/ddp-study-server into PEPPER-754

pegahtah committed:
PEPPER-754
changed the endpoint to work with a single kit instead of array and removed some unnecessary lines

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-754

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,268 @@
+
+    public Object createNonPepperKit(KitRequest kitUploadObject, String studyGuid, String kitTypeName,
+                                    boolean skipAddressValidation, String userIdRequest, AtomicReference<String> shippingCarrier) {
+        DDPInstanceDto ddpInstanceDto = new DDPInstanceDao().getDDPInstanceByStudyGuid(studyGuid).orElseThrow();
```
It looks like some other process adds Juniper studies to the DSM instance table?
How will that work?
As I noted in your design document, I am not comfortable having non-Pepper studies in this table since I don't understand the breadth on the implications.
IOW, it we go this route it would be nice to see an explanation in the design doc about how we keep the studies stored in the DDP instance table from polluting or confusing Pepper.

pegahtah replied:
I think we have discussed this in person before, but let me know if there are any concerns. 

I will add the liquibase for the new Juniper study once we/they know more about the study configs

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,268 @@
+
+        //TODO Pegah fix uploadKit to accept DDPInstanceDto
+        //getting the instance with isHasRole being set to true if the instance has role juniper_study
+        DDPInstance ddpInstance = DDPInstance.getDDPInstanceWithRole(ddpInstanceDto.getInstanceName(), "juniper_study");
```
Should this go above, where we first establish the DDP instance?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,268 @@
+
+
+
+    private void handleKit(@NonNull Connection conn, @NonNull DDPInstance ddpInstance, @NonNull KitType kitType, @NonNull KitRequest kit,
```
What connection is this?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,268 @@
+
+        TransactionWrapper.inTransaction(conn -> {
+            createKit(ddpInstance, kitType, kitUploadObject, kitRequestSettings, easyPostUtil, userIdRequest,
+                    kitTypeName, true, invalidAddressList, duplicateKitList, orderKits,
```
For this code, and the methods below, is there corresponding code for Pepper?
IOW, don't we have to do something just like this for Pepper, or is there something unique about this code?

pegahtah replied:
I now made a method a lot simpler, tuned it to Juniper variables and also combined some methods all together

pegahtah committed:
Merge branch 'develop' into PEPPER-754

pegahtah committed:
PEPPER-754
Cleaned out the code, fixed the code to have more structure and more , Also added safeguards for event trigger and/or writing updates to ES in KitFinalScanUseCase.java, KitFinalSentBaseUseCase.java and KitRequestShipping.java

PEPPER-840
Created ddp/ShipKit Route in JuniperShipKitRoute.java to accept the request and create a new Juniper kit

created KitResponse, KitStatus, ShipKitRequest and JuniperKitRequest and NotPepperKitCreationService.

Both PEPPER-754 and PEPPER-840
added more tests in JuniperKitCreationTest.java

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/DSMServer.java:
```diff
@@ -568,13 +570,15 @@ protected void setupCustomRouting(@NonNull Config cfg) {
-                new JsonTransformer());
+
+        NotPepperKitCreationService notPepperKitCreationService = new NotPepperKitCreationService();
+        get(apiRoot + RoutePath.SHIP_KIT_ENDPOINT, new JuniperShipKitRoute(notPepperKitCreationService), new JsonTransformer());
```
New Route for kit creation

denniscunningham replied:
I realized we are trying to avoid using the term 'Juniper' -- but it is all over this code -- and it looks like we rejected 'external', but it would be really helpful if we came up for a term for these types of kits that would work well for future external kit shipment support. 'NotPepper' feels confusing, but of course that is just my opinion.
Does 'remoteKit' work? I'm up for anything that works, and if we absolutely cannot find a term, I'd prefer to use 'juniper' and remove 'NotPepper'.

denniscunningham replied:
I thought more about this an have an opinion, FWIW. 
The thing that's bothering me are the terms 'notPepper' and 'nonPepper' since they don't fit with the rest of the DSM code. IOW, we don't use the term Pepper in the code (at least not in any prominent places), so calling something non-pepper does not fit with the code.
Could we change 'notPepper' to 'external' throughout this code, and just leave the references to 'Juniper' as is? Does that make sense?

pegahtah replied:
external in terms of DSM vocabulary has a different meaning, and that is when a kit is shipped through or working in a non-broad GP

I am more than happy to change the names to Juniper and keep thinking about a good name. Since it is a first draft, I think we can push this to dev so Juniper team can use it while we come up with a better name. To me, for now, Juniper will do, especially since I don't think we will make this service for anything else. What do you think?

We can also create a ticket to come up with a better name, put the code in dev so Juniper engineers can use it and change the name later

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/DSMServer.java:
```diff
@@ -568,13 +570,15 @@ protected void setupCustomRouting(@NonNull Config cfg) {
+            get(apiRoot + RoutePath.CREATE_CLINICAL_KIT_ENDPOINT, new CreateClinicalDummyKitRoute(new OncHistoryDetailDaoImpl()),
+                    new JsonTransformer());
+            get(apiRoot + RoutePath.CREATE_CLINICAL_KIT_ENDPOINT_WITH_PARTICIPANT, new CreateClinicalDummyKitRoute(new OncHistoryDetailDaoImpl()),
+                    new JsonTransformer());
```
Moving these 2 dummy routes out of production

denniscunningham replied:
These are temporary (i.e. will be removed in the future), or just not ready for production yet?

pegahtah replied:
They are dummy kits created for GP tests last year, We normally take them out of production after the tests are completed.

pegahtah committed:
Merge branch 'develop' into PEPPER-754

pegahtah committed:
PEPPER-754
Moved the check for the study being a Juniper study to L47
Changed the name of NonPepperKitCreationService.addKitRequest method to NonPepperKitCreationService.addJuniperKitRequest

pegahtah committed:
PEPPER-840
Added more static Strings

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -62,17 +32,21 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.IOException;
```
I believe java includes go earlier in the file?
Are you getting any lint feedback?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -996,19 +963,20 @@ public static String writeRequest(@NonNull String instanceId, @NonNull String dd
-                    logger.error(String.format("Error inserting newly created kit request shipping with dsm kit request id: %s in "
-                            + "ElasticSearch", kitRequestShipping.getDsmKitRequestId()));
-                    e.printStackTrace();
+            if (StringUtils.isNotBlank(ddpInstance.getParticipantIndexES())) {
```
I'm guessing this is to exclude pushing Juniper kit info to ES?
If so, maybe a comment explaining that?

pegahtah replied:
Yes, will add

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1084,21 +1052,23 @@ public static void updateKit(long dsmKitId, String userId, DDPInstanceDto ddpIns
             return dbVals;
         });
 
+
```
Maybe remove this extra blank line?

pegahtah replied:
Thank you for pointing it out

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1184,17 +1154,19 @@ public static void updateKit(String dsmKitId, Shipment participantShipment, Ship
         } else {
-            logger.info("Updated kit w/ dsm_kit_id " + dsmKitId, results.resultException);
+            logger.info("Updated kit w/ dsm_kit_id " + dsmKitId);
+            if (ddpInstanceDto.getEsParticipantIndex() != null) {
```
This code looks identical (or perhaps very close to the code above). 
Perhaps create a method to do this?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalScanUseCase.java:
```diff
@@ -62,8 +62,10 @@ protected Optional<ScanError> process(ScanPayload scanPayload) {
                     result = updateKitRequest(kitRequestShipping);
-                    trigerEventsIfSuccessfulKitUpdate(result, ddpLabel, kitRequestShipping);
-                    this.writeSampleSentToES(kitRequestShipping);
+                    if (StringUtils.isNotBlank(kitPayload.getDdpInstanceDto().getEsParticipantIndex())) {
```
This test is repeated in several places.
Perhaps a separate method with a more expressive name would help code readability?
`is<Juniper>Kit`? Where we substitute a better term for 'Juniper'.

pegahtah replied:
Yeah I was thinking about that too, not sure where is best place to put such a method. The check is all over the place

Maybe we can make a code improvement ticket for this epic so we can address them later while getting a working code to dev so we don't block them while we discuss refactoring and design?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1084,21 +1052,23 @@ public static void updateKit(long dsmKitId, String userId, DDPInstanceDto ddpIns
-                logger.error(String.format("Error updating label date of kit request shipping with dsm kit id: %s in ElasticSearch",
-                        dsmKitId));
-                e.printStackTrace();
+            if (StringUtils.isNotBlank(ddpInstanceDto.getEsParticipantIndex())) {
```
See my comment above

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/KitRequest.java:
```diff
@@ -35,6 +35,7 @@ public class KitRequest {
     @DbDateConversion(SqlDateConverter.EPOCH)
     private Long externalOrderDate;
 
+
```
Are the changes in this file necessary?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/KitStatus.java:
```diff
@@ -0,0 +1,7 @@
+package org.broadinstitute.dsm.model.notPepperKit;
+
+public class KitStatus extends KitResponse {
+    public KitStatus(String errorMessage, String juniperKitId) {
```
Why the indirection of KitStatus? IOW, why not just use KitResponse?

pegahtah replied:
This is for now just a placeholder for `KitStatus`, but the reason why I put it like this is, Juniper is thinking of after calling the shipKit for kitCreation, to either get a 400 with an error and a kit id (hence the `KitResponse`) or in case of success, get a 200 with a `KitStatus` message with every field set to either null or info about creation (like created time, etc). They also want to get a `KitStatus` object when they call the status endpoint for getting updates on the kit. Therefore I made the KitStatus as a child to `KitResponse`. This class will have all of the different variables and details of a kit in it

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+@Slf4j
+public class NotPepperKitCreationService {
+
+    public final static String ADDRESS_VALIDATION_ERROR = "UNABLE_TO_VERIFY_ADDRESS";
```
Are these strings codes that the requester is expecting? (Just wondering why they are done this way.)
If so, a comment explaining that would be helpful.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,49 @@
+        }
+        KitResponse kitResponse = this.kitCreationService.createNonPepperKit(juniperKitRequest, shipKitRequest.getJuniperStudyGUID(),
+                shipKitRequest.getKitType());
+        if (kitResponse instanceof KitStatus) {
```
I'm not sure this pattern is particularly obvious. Why not just use KitResponse and check for an error message (or add an error flag if the message is not always an error)?

pegahtah replied:
I see what you mean... 
The reason I made it like that is, when we are returning a `KitStatus` object we can still have errorMessages in it that are not particularly related to this request, like the fedex was out or the cost object didn't work, stuff like that. However, if we had a bad request, we are going to return just a KitResponse with an error message related to that request. I was trying to somehow differentiate based on that.

denniscunningham replied:
At this point KitResponse is identical to KitStatus, but are you imagining that KitStatus will contain more information in the future? If so, who will act on that information?
One way to handle this is "You aren't gonna need it" (YAGNI)... add KitStatus later when it actually is different from KitResponse. But in any case it feels like KitStatus "has a" KitResponse (class composition) rather than KitStatus "is a" KitResponse.
What do you think?

pegahtah replied:
Oh yes, KitStatus will have every tiny detail about the kit. sent date, received date, dsm shipping id, label, who scanned it and when, basically everything. 

Juniper will get it when they call the status endpoint (coming soon :) ) but they want to get the same object for a successful ship kit call as well. I haven't implemented that yet so for the first draft they are ok with just getting a 200 and this needs more consideration and may change in future.

Currently to me, KitStatus _is_ a KitResponse. A KitResponse could be an error, or it could be all of these other informations

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+        if (StringUtils.isBlank(juniperKitRequest.getJuniperKitId())) {
+            return new KitResponse(MISSING_JUNIPER_KIT_ID, null);
+        }
+        Optional<DDPInstanceDto> maybeDdpInstanceDto = new DDPInstanceDao().getDDPInstanceByStudyGuid(studyGuid);
```
Would it be helpful to include the study ID in KitResponse?

pegahtah replied:
Nice! Yes will do!

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+        //getting the instance with isHasRole being set to true if the instance has role juniper_study
+        DDPInstance ddpInstance = DDPInstance.getDDPInstanceWithRole(ddpInstanceDto.getInstanceName(), "juniper_study");
+
+        if (!ddpInstance.isHasRole()) {
```
Are we leaking information by not just saying this is an invalid study ID, as above?
IOW, I'm not sure this provides useful information to a legitimate user, but would provide information to a bad actor.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+            throw new RuntimeException("This is not a Juniper study!");
+        }
+        HashMap<String, KitType> kitTypes = KitType.getKitLookup();
+        String key = kitTypeName + "_" + ddpInstanceDto.getDdpInstanceId();
```
If this is a common pattern for creating the key it should be part of KitType.

pegahtah replied:
That's fair! Will make changes now to the Juniper kit creation part. I'd rather not change the original DSM code until we have great test coverage in place.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+        String key = kitTypeName + "_" + ddpInstanceDto.getDdpInstanceId();
+        KitType kitType = kitTypes.get(key);
+        if (kitType == null) {
+            return new KitResponse(UNKNOWN_KIT_TYPE, juniperKitRequest.getJuniperKitId());
```
As above, should we add kit type to the response object?
Would the requester find that information helpful?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+                KitRequestSettings.getKitRequestSettings(String.valueOf(ddpInstanceDto.getDdpInstanceId()));
+        KitRequestSettings kitRequestSettings = kitRequestSettingsMap.get(kitType.getKitTypeId());
+
+        // if the kit type has sub kits > like for testBoston
```
Unclear what is going on here.

pegahtah replied:
This is the check for subkits (multiple tubes and barcode in one package) . This is not a need right now so I commented the code so we don't spend time debugging it but kept it to for the time that we need it

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+
+
+        TransactionWrapper.inTransaction(conn -> {
+            createKit(ddpInstance, kitType, juniperKitRequest, kitRequestSettings, easyPostUtil, kitTypeName,  orderKits, conn);
```
Not sure what's going on here. How do we know if the kit was created successfully?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+            return null;
+        });
+
+        //send not valid address back to client
```
Is this comment correct?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+     */
+
+    public boolean checkAddress(JuniperKitRequest juniperKitRequest, String phone, EasyPostUtil easyPostUtil) {
+        if ((StringUtils.isBlank(juniperKitRequest.getJuniperParticipantID()) && StringUtils.isBlank(juniperKitRequest.getExternalOrderNumber()))
```
Can you comment on what this test is doing? If we have a participant ID we are ok with not having this other info?
Looks like we use (at least) last name below.

pegahtah replied:
of course, it checks if we have some sort of a kit id (or external id) at this point, and it also checks that there is a last name. It will fail if there is no last name or if there is no id.
However, I will remove the external order check because they don't use external shippers anyways and we don't want to make the mistake of passing the condition with an id that is not used

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+                juniperKitRequest.setEasypostAddressId(address.getId());
+                return true;
+            } catch (EasyPostException e) {
+                log.error("Easypost couldn't create an address for " + juniperKitRequest.getShortId());
```
Is it possible this was just a bad request?
If so we should probably not log at the error level.

pegahtah replied:
changed to warn

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+        //In case it's needed, subkits handling should be added here
+        String shippingId;
+        String userId;
+        //checking ddpInstance.isHasRole() to know this is a Juniper Kit
```
It seems like we already did this checking elsewhere.
If so, perhaps something more like an assertion at the beginning of the method is needed (a check that throws a 500 error).

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+            shippingId = JUNIPER_UNDERSCORE + kit.getJuniperKitId();
+            userId = JUNIPER;
+        } else {
+            log.error("Seems like {} is not a JUNIPER study! ", ddpInstance.getName());
```
If this is a request error we should not log at error level.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+            if (kitRequestSettings.getCollaboratorSampleTypeOverwrite() != null) {
+                bspCollaboratorSampleType = kitRequestSettings.getCollaboratorSampleTypeOverwrite();
+            }
+            if (StringUtils.isNotBlank(collaboratorParticipantId)) {
```
Perhaps a comment about what the collaborator participant ID indicates?

pegahtah replied:
Not sure how or what to add, it is the id of the participant in GP's view... (I'll add that but not sure how much value it adds to the code reader, unfortunately some of these are just values and ids that GP needs and may not make any sense outside of the scope of DSM)

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/notPepperKit/NotPepperKitCreationService.java:
```diff
@@ -0,0 +1,238 @@
+    /**
+     * This method creates a collaborator participant id for the kit and then inserts in DB
+     */
+    private void createKit(@NonNull DDPInstance ddpInstance, @NonNull KitType kitType, JuniperKitRequest kit,
```
Somewhere here or in the caller(s) there should be some exception handling?

pegahtah committed:
PEPPER-840
renamed to NonPepperKitCreationService
removed extra DDPInstanceDto object in NonPepperKitCreationService to avoid extra call
Added a new method, getDDPInstanceWithRoleByStudyGuid, in DDPInstance.java

pegahtah committed:
PEPPER-840
checkstyle fixes

pegahtah committed:
PEPPER-840
removing a blank line for checkstyle fixes,
also refactoring the upsert in KitRequestShipping.java to a method KitRequestShipping.upsertUpdatedKitInfoIntoES

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -996,19 +991,21 @@ public static String writeRequest(@NonNull String instanceId, @NonNull String dd
+                } catch (Exception e) {
+                    //This error will trigger on studies with no participants, this skips
+                    //the error log if that is the reason for the upsert failure.
+                    if (StringUtils.isNotBlank((ddpInstance.getParticipantIndexES()))) {
```
It seems like we did this check above to enter this code. Has anything changed since then?

pegahtah replied:
removed

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1091,15 +1088,7 @@ public static void updateKit(long dsmKitId, String userId, DDPInstanceDto ddpIns
-                        dsmKitId));
-                e.printStackTrace();
-            }
+            upsertUpdatedKitInfoIntoES(ddpInstanceDto, kitRequestShipping, dsmKitId);
```
Are we calling upsertUpdatedKitInfoIntoES for Juniper kits? If so, why?

pegahtah replied:
I added the check in the method, so that method has a check if there is an index in ES

pegahtah committed:
Merge branch 'develop' into PEPPER-754

pegahtah committed:
PEPPER-840
added the value received by DSM for studyGuid to the KitResponse

pegahtah committed:
PEPPER-840
* added more assertions to JuniperKitCreationTest
* added a value field for KitResponse
* added KitType.createKitTypeKey to refactor how the key is made, haven't made changes in the original DSM code yet
* applied PR comments to the NonPepperKitCreationService, specifically added a DSM_ERROR, caught the exception in createKit method, added more comments and changed some of the errors to warnings,

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-754

pegahtah committed:
PEPPER-840
removed extra unnecessary check

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1184,10 +1171,16 @@ public static void updateKit(String dsmKitId, Shipment participantShipment, Ship
+    }
 
+    private static void upsertUpdatedKitInfoIntoES(DDPInstanceDto ddpInstanceDto, KitRequestShipping kitRequestShipping, Object dsmKitId) {
+        if (ddpInstanceDto.getEsParticipantIndex() != null) {
```
@denniscunningham this prevents updates for Juniper

denniscunningham replied:
Right. As we discussed it would be nice to create a more explicit way of expressing that.

pegahtah replied:
I moved it to DDPInstanceDto and DDPInstance classes, since it is not really showing a Juniper kit , but the ability for the study, wether or not it has the index to update

pegahtah committed:
PEPPER-840
removed unnecessary logging found in JWTRouteFilter
changed JuniperShipKitRoute.java to implement Route
changed endpoint to POST

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -0,0 +1,276 @@
+        KitRequestSettings kitRequestSettings = kitRequestSettingsMap.get(kitType.getKitTypeId());
+
+        // if the kit type has sub kits > like for testBoston
+        //        boolean kitHasSubKits = kitRequestSettings.getHasSubKits() != 0;
```
Is this code still relevant?

pegahtah replied:
I want to keep the subkit code in there so in any case if anyone wants to add them in neaqr future there is a clue of where to begin

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -0,0 +1,276 @@
+                juniperKitRequest.setEasypostAddressId(address.getId());
+                return true;
+            } catch (EasyPostException e) {
+                log.warn("Easypost couldn't create an address for " + juniperKitRequest.getShortId());
```
Unclear what's going on here. Can you add a comment?
Will we get this exception for a bad request, or is this more like an internal error?
Why are we logging at warning level?
The results of the stack trace are going, I assume, to stdout or stderr, which may or may not be logged.

pegahtah replied:
addressed

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -0,0 +1,276 @@
+        ArrayList<KitRequest> orderKits = new ArrayList<>();
+
+
+        SimpleResult result = TransactionWrapper.inTransaction(conn -> {
```
Is there a reason this transaction clause is not in `createKit`?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -0,0 +1,276 @@
+        String juniperKitRequestId;
+        String userId;
+        //checking ddpInstance.isHasRole() to know this is a Juniper Kit
+        if (StringUtils.isNotBlank(kit.getJuniperKitId()) || ddpInstance.isHasRole()) {
```
It seems like we already checked this further up the stack?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -0,0 +1,276 @@
+                                              int kitTypeId, String collaboratorParticipantId, String errorMessage,
+                                              EasyPostUtil easyPostUtil,
+                                              JuniperKitRequest kit, String externalOrderNumber, String juniperKitRequestId,
+                                              String ddpLabel, String userId, SimpleResult transactionResults) {
```
I'm not a fan of lots of parameters -- it makes for poor readability. Is there a way to encapsulate these into a class that is acting as a struct? Or more conventionally, create a kit shipping class that has much of this info as instance variables.

pegahtah replied:
I agree, we can do this in the next versions

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -0,0 +1,276 @@
+            throw new RuntimeException("EasyPost addressId could not be received ", e);
+        }
+
+        if (StringUtils.isNotBlank(kitRequestSettings.getExternalShipper())) {
```
Can you add a comment about how we handle external shippers differently?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -0,0 +1,276 @@
+            if (kitRequestSettings.getCollaboratorSampleTypeOverwrite() != null) {
+                bspCollaboratorSampleType = kitRequestSettings.getCollaboratorSampleTypeOverwrite();
+            }
+            if (StringUtils.isNotBlank(collaboratorParticipantId)) {
```
Can you comment on what this case is? What is a `collaboratorParticipantId` and what does this code accomplish?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -0,0 +1,276 @@
+            if (StringUtils.isNotBlank(collaboratorParticipantId)) {
+                collaboratorSampleId =
+                        KitRequestShipping.generateBspSampleID(conn, collaboratorParticipantId, bspCollaboratorSampleType, kitTypeId);
+                if (collaboratorParticipantId == null) {
```
How can this be null? The test above (isNotBlank) should assure that?

pegahtah replied:
true! hmmm! the code is like this in the KitUpload and KitRequestShipping and I just moved it over... I will remove this check

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -0,0 +1,276 @@
+                    errorMessage += "collaboratorParticipantId was too long ";
+                }
+                if (collaboratorSampleId == null) {
+                    errorMessage += "collaboratorSampleId was too long ";
```
I don't think this works (this will not update errorMessage for the caller). Do you have a test for this?
Also, it seems like you are attempting to return errors in two different ways, via the return and via this parameter. One way would be less confusing.

pegahtah replied:
No the error message here will be written to the database. It is one of the GP errors if we can't fit a participant id into their desired length for barcodes

denniscunningham replied:
Oh, ok. So this is a non-fatal error? Does GP view it? If so, how?

I'm unclear on why errorMessage is passed into addJuniperKitRequest. It appears to be empty when passed.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -0,0 +1,276 @@
+                                kit.getExternalOrderNumber(),
+                                false,
+                                null, ddpInstance, bspCollaboratorSampleType, ddpLabel);
+                kit.setDdpLabel(juniperKitRequestId);
```
This clause is similar to that above, but here it includes `kit.setDdpLabel`. Is that intentional?

pegahtah replied:
it is not used anymore so I removed it

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/JuniperKitStatus.java:
```diff
@@ -0,0 +1,7 @@
+package org.broadinstitute.dsm.model.nonpepperkit;
+
+public class JuniperKitStatus extends KitResponse {
```
I still don't think this provides a clear way to implement response handling. IOW, I do not see what value this technique provides.

pegahtah replied:
I don't think there is any value one way or the other. IOW it's just a matter of how you look at if JuniperKitStatus is a KitResponse or has a KitResponse

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -0,0 +1,276 @@
+     *
+     * @return SimpleResult of the interaction with database
+     */
+    private SimpleResult createKit(@NonNull DDPInstance ddpInstance, @NonNull KitType kitType, JuniperKitRequest kit,
```
The use of SimpleResult as a return type feels clumsy. We are jamming response information into a throwable type that we ultimately will not throw.
Would it make more sense to return a KitResponse?

pegahtah committed:
PEPPER-840
PR comments

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -124,8 +124,8 @@ public boolean checkAddress(JuniperKitRequest juniperKitRequest, String phone, E
-                log.warn("Easypost couldn't create an address for " + juniperKitRequest.getShortId());
-                e.printStackTrace();
+                // log the reason for address creation failure and return false. The method will then return the error code
+                log.warn("Easypost couldn't create an address for " + juniperKitRequest.getShortId(), e);
```
Is this due to an bad request or an internal error?

denniscunningham reviewed (approved):
I don't think this code is ready for PROD, so my approval is only for release to Dev so Juniper can test. 
We need to figure out how to limit the impact of this code should it not have the desired changes before the next DSM release.

pegahtah committed:
PEPPER-840
Adding ignore to tests
